diff --git a/python_configure.bzl b/python_configure.bzl
index 1f5bffa..89e0542 100644
--- a/python_configure.bzl
+++ b/python_configure.bzl
@@ -251,9 +252,19 @@ def _get_python_include(repository_ctx, python_bin):
         [
             python_bin,
             "-c",
-            "from __future__ import print_function;" +
-            "from distutils import sysconfig;" +
-            "print(sysconfig.get_python_inc())",
+            """
+from __future__ import print_function
+from distutils import sysconfig
+
+import shutil
+
+python_inc_dir = sysconfig.get_python_inc()
+config_h_path = sysconfig.get_config_h_filename()
+
+shutil.copyfile(config_h_path, python_inc_dir + "/pyconfig.h")
+
+print(python_inc_dir)
+            """,
         ],
         error_msg = "Problem getting python include path.",
         error_details = ("Is the Python binary path set up right? " +
